drivers/serial/uart_mcux_flexcomm.c

```diff
@@ -342,6 +379,636 @@ static int mcux_flexcomm_uart_config_get(const struct device *dev,
+
+	if (!device_is_ready(config->rx_dma.dev) ||
+		!device_is_ready(config->tx_dma.dev)) {
+		return -ENODEV;
+	}
+
+	/* Disable DMA requests */
+	USART_EnableTxDMA(config->base, false);
+	USART_EnableRxDMA(config->base, false);
+
+	/* Route DMA requests */
+#if defined(CONFIG_SOC_SERIES_IMX_RT5XX) || defined(CONFIG_SOC_SERIES_IMX_RT6XX)
+	/* RT 3 digit uses input mux to route DMA requests from
+	 * the UART peripheral to a hardware designated DMA channel
+	 */
+	INPUTMUX_Init(INPUTMUX);
+	INPUTMUX_EnableSignal(INPUTMUX,
+		fc_uart_calc_inmux_connection(config->rx_dma.channel,
+					config->rx_dma.base), true);
+	INPUTMUX_EnableSignal(INPUTMUX,
+		fc_uart_calc_inmux_connection(config->tx_dma.channel,
+					config->tx_dma.base), true);
+	INPUTMUX_Deinit(INPUTMUX);
+#endif /* RT5xx and RT6xx */
+
+	/* Init work objects for RX and TX timeouts */
+	k_work_init_delayable(&data->tx_data.timeout_work,
+			config->tx_timeout_func);
+	k_work_init_delayable(&data->rx_data.timeout_work,
+			config->rx_timeout_func);
+
+	/* TX FIFO interrupt; configure to interrupt when empty */
+	config->base->FIFOTRIG &= ~USART_FIFOSTAT_TXLVL_MASK;
+
+	return 0;
+}
+
+#endif /* CONFIG_UART_ASYNC_API */
+
+
+static void mcux_flexcomm_isr(const struct device *dev)
+{
+	struct mcux_flexcomm_data *data = dev->data;
+
+#ifdef CONFIG_UART_INTERRUPT_DRIVEN
+	if (data->irq_callback) {
+		data->irq_callback(dev, data->irq_cb_data);
+	}
+#endif
+
+#ifdef CONFIG_UART_ASYNC_API
+	const struct mcux_flexcomm_config *config = dev->config;
+
+	/* If there is an async callback then we are using async api */
+	if (data->async_callback) {
+
+		/* Handle RX interrupt (START bit detected)
+		 * RX interrupt defeats the purpose of UART ASYNC API
+		 * because core is involved for every byte but
+		 * it is included for compatibility of applications.
+		 * There is no other way with flexcomm UART to handle
+		 * Zephyr's RX ASYNC API. However, if not using the RX
+		 * timeout (timeout is forever), then the performance is
+		 * still as might be expected.
+		 */
+		if (config->base->INTSTAT & USART_INTSTAT_START_MASK
+			&& data->async_callback) {
+
+			/* Receiving some data so reschedule timeout,
+			 * unless timeout is 0 in which case just handle
+			 * rx data now. If timeout is forever, don't do anything.
+			 */
+			if (data->rx_data.timeout == 0) {
+				flexcomm_uart_rx_update(dev);
+			} else if (data->rx_data.timeout != SYS_FOREVER_US) {
+				k_work_reschedule(&data->rx_data.timeout_work,
+						K_USEC(data->rx_data.timeout));
+			}
+
+			/* W1C start bit status bit */
+			config->base->STAT |= 0b1 & USART_STAT_START_MASK;
+		}
+
+		/* Handle TX interrupt (TXLVL = 0)
+		 * TX interrupt should happen when transfer is complete
+		 * because DMA filling TX fifo is faster than transmitter rate
+		 */
+		if (config->base->FIFOINTSTAT & USART_FIFOINTSTAT_TXLVL_MASK
+			&& data->async_callback) {
+
+			/* Disable interrupt */
+			config->base->FIFOINTENCLR |= USART_FIFOINTENCLR_TXLVL_MASK;
```

Comment:
Do we need `|=` here? This looks like a register where we can just write directly, as it seems all the bits in are read as zero

Reply:
I think it would look strange/confusing in the code to have just `=` when writing to a register to clear a specific bit, this is the common practice

Reply:
Writing directly to a `SET` register is common practice- it avoids an unnecessary register read, and is one of the reasons that a `SET` register exists at all.
```
grep -r "_SET = ." --include="*.c"
drivers/timer/mchp_xec_rtos_timer.c:    ECIA_XEC_REGS->GIRQ[girq - 8].EN_SET = BIT(bitpos);
drivers/interrupt_controller/intc_mchp_ecia_xec.c:              regs->BLK_EN_SET = BIT(girq_num);
drivers/interrupt_controller/intc_mchp_ecia_xec.c:      regs->GIRQ[girq_num - MCHP_FIRST_GIRQ].EN_SET = BIT(src_bit_pos);
drivers/interrupt_controller/intc_mchp_ecia_xec.c:      regs->GIRQ[girq_num - MCHP_FIRST_GIRQ].EN_SET = bitmap;
drivers/interrupt_controller/intc_mchp_ecia_xec.c:      regs->GIRQ[girq - MCHP_FIRST_GIRQ].EN_SET = BIT(src);
drivers/interrupt_controller/intc_mchp_ecia_xec.c:      ecia->BLK_EN_SET = MCHP_ECIA_AGGR_BITMAP;
soc/arm/microchip_mec/mec1501/device_power.c:   GIRQ22_REGS->EN_SET = (1ul << 9);
soc/arm/microchip_mec/mec1501/soc.c:    ECIA_REGS->BLK_EN_SET = MCHP_ECIA_AGGR_BITMAP;
soc/arm/microchip_mec/mec172x/device_power.c:   regs->GIRQ22.EN_SET = MCHP_ESPI_WK_CLK_GIRQ_BIT;
soc/arm/microchip_mec/mec172x/device_power.c:   regs->GIRQ21.EN_SET = MCHP_KEYSCAN_GIRQ_BIT;
soc/arm/microchip_mec/mec172x/device_power.c:   regs->GIRQ21.EN_SET = MCHP_PS2_0_PORT0B_WK_GIRQ_BIT;
soc/arm/nxp_imx/rt/power_rt10xx.c:              XTALOSC24M->LOWPWR_CTRL_SET = XTALOSC24M_LOWPWR_CTRL_LPBG_SEL_MASK;
soc/arm/nxp_imx/rt/power_rt10xx.c:              PMU->MISC0_SET = PMU_MISC0_REFTOP_PWD_MASK;
soc/arm/nxp_imx/rt/lpm_rt1064.c:        CCM_ANALOG->PLL_USB1_SET = CCM_ANALOG_PLL_USB1_BYPASS_MASK;
soc/arm/nxp_imx/rt/lpm_rt1064.c:        CCM_ANALOG->PLL_USB1_SET = CCM_ANALOG_PLL_USB1_ENABLE_MASK;
soc/arm/nxp_imx/rt/soc_rt11xx.c:        CCM->GPR_PRIVATE1_SET = 0x1;
soc/arm/nxp_imx/rt5xx/flash_clock_setup.c:      CLKCTL0->PSCCTL0_SET = CLKCTL0_PSCCTL0_SET_FLEXSPI0_OTFAD_CLK_MASK;
soc/arm/nxp_imx/rt5xx/flash_clock_setup.c:                      CLKCTL0->PSCCTL0_SET = CLKCTL0_PSCCTL0_SET_FLEXSPI0_OTFAD_CLK_MASK;
soc/arm/nxp_imx/rt5xx/flash_clock_setup.c:                      CLKCTL0->PSCCTL0_SET = CLKCTL0_PSCCTL0_SET_FLEXSPI1_CLK_MASK;
soc/arm/nxp_imx/rt6xx/flash_clock_setup.c:                      CLKCTL0->PSCCTL0_SET = CLKCTL0_PSCCTL0_SET_FLEXSPI_OTFAD_CLK_MASK;
```

Reply:
ok I guess the same is true for the set register then too?